Lib/shutil.py

```diff
@@ -81,6 +94,104 @@ def copyfileobj(fsrc, fdst, length=16*1024):
+    """
+    try:
+        infd = fsrc.fileno()
+        outfd = fdst.fileno()
+    except Exception as err:
+        raise _GiveupOnZeroCopy(err)  # not a regular file
+
+    try:
+        posix._fcopyfile(infd, outfd)
+    except OSError as err:
+        if err.errno in {errno.EINVAL, errno.ENOTSUP}:
+            raise _GiveupOnZeroCopy(err)
+        else:
+            raise err from None
+
+def _zerocopy_win(fsrc, fdst):
+    """Copy 2 files by using high-performance CopyFileW (Windows only)."""
+    nt._win32copyfile(fsrc, fdst)
+
+def _zerocopy_sendfile(fsrc, fdst):
+    """Copy data from one regular mmap-like fd to another by using
+    high-performance sendfile() method.
+    This should work on Linux >= 2.6.33 and Solaris only.
+    """
+    global _HAS_SENDFILE
+    try:
+        infd = fsrc.fileno()
+        outfd = fdst.fileno()
+    except Exception as err:
+        raise _GiveupOnZeroCopy(err)  # not a regular file
+
+    # Hopefully the whole file will be copied in a single call.
+    # sendfile() is called in a loop 'till EOF is reached (0 return)
+    # so a bufsize smaller or bigger than the actual file size
+    # should not make any difference, also in case the file content
+    # changes while being copied.
+    try:
+        blocksize = max(os.fstat(infd).st_size, 2 ** 23)  # min 8MB
+    except Exception:
+        blocksize = 2 ** 27  # 128MB
+
+    offset = 0
+    while True:
+        try:
+            sent = os.sendfile(outfd, infd, offset, blocksize)
+        except OSError as err:
+            if err.errno == errno.ENOTSOCK:
+                # sendfile() on this platform (probably Linux < 2.6.33)
+                # does not support copies between regular files (only
+                # sockets).
+                _HAS_SENDFILE = False
+                raise _GiveupOnZeroCopy(err)
+
+            if err.errno == errno.ENOSPC:  # filesystem is full
+                raise err from None
+
+            # Give up on first call and if no data was copied.
+            if offset == 0 and os.lseek(outfd, 0, os.SEEK_CUR) == 0:
+                raise _GiveupOnZeroCopy(err)
+
+            raise err from None
+        else:
+            if sent == 0:
+                break  # EOF
+            offset += sent
+
+def _copyfileobj2(fsrc, fdst):
+    """Copy 2 regular mmap-like fds by using zero-copy sendfile(2)
+    (Linux) and fcopyfile(2) (OSX) syscalls.
+    In case of error fallback on using plain read()/write() if no
+    data was copied.
+    """
+    # Note: copyfileobj() is left alone in order to not introduce any
+    # unexpected breakage. Possible risks by using zero-copy calls
+    # in copyfileobj() are:
+    # - fdst cannot be open in "a"(ppend) mode
+    # - fsrc and fdst may be open in "t"(ext) mode
+    # - fsrc may be a BufferedReader (which hides unread data in a buffer),
+    #   GzipFile (which decompresses data), HTTPResponse (which decodes
+    #   chunks).
+    # - possibly others
+    if _HAS_SENDFILE:
```

Comment:
Given a configure check for `_HAS_SENDFILE`, the initial check for both it and `_HAS_FCOPYFILE` can be moved up to the module level, and define `_copyfileobj_fast` differently based on the results. If both flags are false, then `_copyfileobj_fast` can be aliased directly to `copyfileobj`. (I believe Mac OS X defines both operations, so it will be important to give `_HAS_FCOPYFILE` precedence when deciding which implementation of `_copyfileobj_fast` to use)

The dynamic check to handle the case where a Python built on a newer distro is run against an older kernel (and hence still gets `ENOTSOCK` at runtime) can then also be handled by setting `_copyfileobj_fast = copyfileobj` rather than by setting a flag that needs to be checked on every call.